### ISSUE TITLE
Model predict returns result of different quality then validation declares.

### ISSUE THREAD

**Issue by rybickibartek** [Microsoft/LightGBM#1692]
## Environment info

Operating System: Fedora 24

CPU: Intel(R) Core(TM) i5-7200U CPU @ 2.50GHz

C++/Python/R version: Python 3.6.6

Conda env:

  - python=3.6
  - numpy=1.15
  - scikit-learn=0.19.1
  - pandas=0.23.3
  - matplotlib=2.2.2
  - jupyter
  - jupyterlab
  - pip:
    - sklearn-pandas==1.6.0
    - featuretools==0.2.1
    - lightgbm==2.2.0
    - tqdm

## Problem description

I create train and valid datasets and run training with this two datasets. All the values are int, float or nan. Some of the columns are categorical [encoded by LabelEncoder]. 

When I run training I got some valid's rmse and best iteration number. Then I call a predict method on the valid dataframe and I got ... different rmse value. 

Am I using categorical features in a not correct way? When I treat all the features as a non-categorical features (I mean delete all three 'categorical_feature=cat_cols' from the code) then problem do NOT exists. I try different boosting methods but problem also exists.

I the below example difference is around 0.006 but I also have an examples in which the gap is much bigger around 0.05.

## Reproducible examples

        params = {"objective" : "regression", "metric" : "l2_root","verbose" : -1}

        train_data = lgb.Dataset(data=X_train,
                                 label=y_train,
                                 categorical_feature=cat_cols,
                                 free_raw_data=False)

        valid_data = lgb.Dataset(data=X_valid,
                                 label=y_valid,
                                 categorical_feature=cat_cols,
                                 free_raw_data=False)

        model = lgb.train(train_set = train_data,
                          params = params,
                          num_boost_round = NUM_ROUNDS,
                          early_stopping_rounds = STOP_ROUNDS,
                          categorical_feature=cat_cols,
                          valid_sets=[train_data, valid_data],
                          verbose_eval = VERBOSE_EVAL)

        print('best iter:', model.best_iteration)

        y_valid_pred = model.predict(X_valid)
        print(mean_squared_error(y_valid, y_valid_pred) ** 0.5)

## Output
Early stopping, best iteration is:
[97]	training's rmse: 1.43167	valid_1's rmse: 1.68**23**

best iter: 97
1.68**82**881181967648

**Comment by guolinke** [comment]
@rybickibartek can you try "reg_sqrt=false" ?

**Comment by rybickibartek** [comment]
To be honest when I restart my computer problem does not occur. But I will try req_sort = false in the case of the same problem, and give you a feedback about the effect.

**Comment by StrikerRUS** [comment]
Feel free to reopen if needed.
